### PR TITLE
TRA-4602_timeline bar charts to traffic statistics

### PR DESCRIPTION
**Pull request by AmitUp9**
Screen Shot 2022-06-21 at 14 04 15

### CODE REVIEW

File: ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.module.sass [hunk 1]
```diff
@@ -0,0 +1,5 @@
+.barChartContainer
+    width: 100%
+    display: flex 
+    justify-content: center
+
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.module.sass:5]: new line

File: ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    const creatUniqueObjArray = (objArray) => {
+        return [
+            ...new Map(objArray.map((item) => [item["name"], item])).values(),
+        ];
+    }
+
+    useEffect(() => {
+        if (!data) return;
+        let protocolsBarsData = [];
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx:42]: you can use const - you can push new items into a const array

File: ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        let prtcNames = [];
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx:43]: you can use const - you can push new items into a const array

File: ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        data.map(protocolObj => {
+            let obj: { [k: string]: any } = {};
+            obj.timestamp = getHoursAndMinutes(protocolObj.timestamp);
+            protocolObj.protocols.map(protocol => {
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx:47]: its better to use foreach if you don't use he returned array of map

File: ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                obj[`${protocol.name}`] = protocol[StatsMode[timeLineBarChartMode]];
+                prtcNames.push({name: protocol.name, color: protocol.color});
+            })
+            protocolsBarsData.push(obj);
+        })
+        let uniqueObjArray = creatUniqueObjArray(prtcNames);
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx:53]: const

File: ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        protocolsBarsData.sort((a, b) => a.timestamp < b.timestamp ? -1 : 1);
+        setProtocolStats(protocolsBarsData);
+        setProtocolsNamesAndColors(uniqueObjArray);
+    }, [data,timeLineBarChartMode])
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx:57]: add all the setters and the functions you use here as dependencies

File: ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+    const bars = useMemo(() => protocolsNamesAndColors.map((protoclToDIsplay) => {
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx:59]: protoclToDIsplay => protocolToDisplay

File: ui-common/src/components/modals/TrafficStatsModal/TrafficStatsModal.tsx [hunk 8]
```diff
@@ -19,39 +20,44 @@ const modalStyle = {
...
-    if(isOpen && getTrafficStatsDataApi) {
+    if (isOpen && getPieStatsDataApi) {
       (async () => {
         try {
           setIsLoading(true);
-          const data = await getTrafficStatsDataApi();
-          setStatsData(data);
+          const pieData = await getPieStatsDataApi();
+          setPieStatsData(pieData);
+          const timelineDta = await getTimelineStatsDataApi();
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/TrafficStatsModal/TrafficStatsModal.tsx:51]: Dta => Data

File: ui-common/src/components/modals/TrafficStatsModal/TrafficStatsModal.tsx [hunk 9]
```diff
@@ -19,39 +20,44 @@ const modalStyle = {
...
+          setTimelineStatsData(timelineDta);
         } catch (e) {
           console.error(e)
         } finally {
           setIsLoading(false)
         }
       })()
     }
-  }, [isOpen, getTrafficStatsDataApi])
+  }, [isOpen, getPieStatsDataApi])
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/TrafficStatsModal/TrafficStatsModal.tsx:60]: you miss some dependencies here

File: ui-common/src/components/modals/TrafficStatsModal/TrafficStatsModal.tsx [hunk 10]
```diff
@@ -65,19 +71,24 @@ export const TrafficStatsModal: React.FC<TrafficStatsModalProps> = ({ isOpen, on
...
               <select className={styles.select} value={statsMode} onChange={(e) => setStatsMode(e.target.value)}>
-                {modes.map(mode => <option value={mode}>{mode}</option>)}
+                {modes.map(mode => <option key={mode} value={mode}>{mode}</option>)}
               </select>
             </div>
-            {isLoading ? <div style={{textAlign: "center", marginTop: 20}}>
-                <img alt="spinner" src={spinnerImg} style={{ height: 50 }} />
-            </div> : <TrafficPieChart pieChartMode={statsMode} data={statsData}/>}
+            <div>
+            {isLoading ? <div style={{ textAlign: "center", marginTop: 20 }}>
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/TrafficStatsModal/TrafficStatsModal.tsx:85]: For not displaying the spinner twice you can use:
{isLoading && <spinner...> : <React.Fragment><div>pie</div><div>bar</div></React.Fragment>}

**Review by lirazyehezkel — COMMENTED** [review]
added some comments

File: ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.module.sass (hunk 1 above)
Comment (AmitUp9) [reply ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.module.sass:5]: Its suppose to have one line gap isn't it?

File: ui-common/src/components/modals/TrafficStatsModal/TrafficStatsModal.tsx [hunk 11]
```diff
@@ -82,12 +81,14 @@ export const TrafficStatsModal: React.FC<TrafficStatsModalProps> = ({ isOpen, on
...
             <div>
-            {isLoading ? <div style={{ textAlign: "center", marginTop: 20 }}>
-              <img alt="spinner" src={spinnerImg} style={{ height: 50 }} />
-            </div> : <div><TrafficPieChart pieChartMode={statsMode} data={pieStatsData} /> </div>}
-            {isLoading ? <div style={{ textAlign: "center", marginTop: 20 }}>
-              <img alt="spinner" src={spinnerImg} style={{ height: 50 }} />
-            </div> : <div> <TimelineBarChart timeLineBarChartMode={statsMode} data={timelineStatsData} /> </div>}
+              {isLoading ? <div style={{ textAlign: "center", marginTop: 20 }}>
+                <img alt="spinner" src={spinnerImg} style={{ height: 50 }} />
+              </div> : <React.Fragment>
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/TrafficStatsModal/TrafficStatsModal.tsx:86]: you actually don't need the react fragment if you have common div container

**Review by lirazyehezkel — APPROVED** [review]

File: ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    timeLineBarChartMode: string;
+    data: any;
+}
+
+export const TimelineBarChart: React.FC<TimelineBarChartProps> = ({ timeLineBarChartMode, data }) => {
+    const [protocolStats, setProtocolStats] = useState([]);
+    const [protocolsNamesAndColors, setProtocolsNamesAndColors] = useState([]);
+
+    const padTo2Digits = useCallback((num) => {
+        return String(num).padStart(2, '0');
```
Comment (leon-up9) [inline ui-common/src/components/modals/TrafficStatsModal/TimelineBarChart/TimelineBarChart.tsx:24]: maybe should be a utility function?

**Pull request merged**